src/splinelib_ext/nurbspl.hpp

```diff
@@ -139,6 +217,378 @@ class NurbsExt : public splinelib::sources::splines::Nurbs<para_dim, dim> {
+    clipped.fill(0);
+    prevclipped.fill(0);
+    solverclip.fill(0);
+    
+    
+    // start with some sort of guess
+    // TODO: better option than this?
+    _good_guess(query,
+                guessoption,
+                current_guess,
+                searchbounds);
+
+    constexpr int max_iter = para_dim * 20; /* max newton iter */
+
+    double prevnorm{123456789.}, curnorm;
+
+    /* newton loops */
+    for (int i{0}; i < max_iter; i++) {
+      /* build system to solve*/
+      _distance(current_guess, query, dist);
+      _build_djdu(current_guess, dist, eyeders, djdu); /* rhs */
+      _build_d2jdu2(current_guess, dist, eyeders, d2jdu2); /* lhs */
+
+      /* debug prints */
+      std::cout << "**************** Newton iteration: " << i << " **********"; 
+      std::cout << "\n distance (goal - spline(para_c): ";
+      for (int i{0}; i < para_dim; i++) {
+        std::cout << dist[i] << " ";
+      }
+      std::cout << "\n";
+      std::cout << "System Matrix: A | b  (d2jdu2 | djdu)\n";
+      for (int i{0}; i < para_dim; i++) {
+        for (int j{0}; j < para_dim; j++) {
+          std::cout << d2jdu2[i][j] << " ";
+        }
+        std::cout << "| " << djdu[i];
+        std::cout << "\n";
+      }
+      /* end debug prints */
+
+      /* solve and update */
+      // solver clip only if it is clipped at the same place twice.
+      if(prevclipped == clipped && !solverclipped) {
+        solverclip = clipped;
+        solverclipped = true;
+      }
+      // solve. pivoting is done inplace, so don't use these afterwards.
+      // dx is always reorganized.
+      // TODO: why not reorganize d2jdu2 and djdu too?
+      gauss_with_pivot(d2jdu2, djdu, solverclip, dx); /* solve */
+      ew_iplus(dx, current_guess); /* update */
+
+
+      /* debug prints */
+      std::cout << "Solution (dx): ";
+      for (int i{0}; i < para_dim; i++) {
+        std::cout << dx[i] << " ";
+      }
+      std::cout << "\nUpdated: ";
+      for (int i{0}; i < para_dim; i++) {
+        std::cout << current_guess[i].Get() << " ";
+      }
+      /* end debug prints */
+
+      /* clip */
+      prevclipped = clipped;
+      clip(searchbounds, current_guess, clipped); /* clip */
+
+
+      /* debug prints */
+      std::cout << "\nClip info: ";
+      for (int i{0}; i < para_dim; i++) {
+        std::cout << clipped[i] << " ";
+      }
+      std::cout  << "\n Current guess after clip:";
+      for (int i{0}; i < para_dim; i++) {
+        std::cout << current_guess[i].Get() << " ";
+      }
+      /* end debug prints */
+
+      // Customer satisfaction survey
+      // ============================
+      // 1. Converged?
+      //   -> should be zero, if query point is on the spline.
+      // 2. All clipped?
+      //   -> should be the case if query point lies outside the spline.
+      curnorm = norm2(djdu);
+      std::cout  << "\n norm2(djdu): " << curnorm << "\n";
+      if (std::abs(prevnorm - curnorm) < tolerance
```

Comment:
Will never be fulfilled on first iteration. Although the first iteration can be sufficient. If